Python/Product/Analysis/LanguageServer/SingleThreadSynchronizationContext.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer {
```

Comment:
Since this isn't really language server specific, it could go into Infrastructure.

Reply:
Or it can stay in the VSCode LS host?

Reply:
This is to use Post and have PublishDiagnostic to be callable under the lock since it fires events.

Reply:
Is it worth having this on Server itself then? Then we can route all events through it automatically

Reply:
I can move it to the server base then

Reply:
On the other hand, it is not strictly needed there. Editor file uses Post to execute sequence of code or events. Routing events at the server level still may cause mix ups. Currently loop is executed fully at once, 
```cs
            _syncContext.Post(_ => {
                foreach (var a in args) {
```
but if Post becomes part of `_server.PublishDiagnostics` it will no longer hold true.